### PR TITLE
Shard key - create collection

### PR DESCRIPTION
**Pull request by generall**
Tracking issue: https://github.com/qdrant/qdrant/issues/2807

- Add params to create collection with custom sharding
- fix snapshot persistence - save shard mapping with collection snapshot
- allow loading of shards by list of ids instead of iterator

### CODE REVIEW

File: lib/storage/src/content_manager/toc/create_collection.rs [hunk 1]
```diff
@@ -66,13 +69,28 @@ impl TableOfContent {
...
+                    shard_number
+                } else {
+                    collection_shard_distribution.shard_count() as u32
+                }
+            }
+            ShardingMethod::Custom => {
+                if let Some(shard_number) = shard_number {
+                    shard_number
+                } else {
+                    default_shard_number().get()
```
[the remark this answers is not in the document]
Comment (generall) [reply lib/storage/src/content_manager/toc/create_collection.rs:89]: No, default shard number of custom sharding is 1.

File: lib/collection/src/collection/snapshots.rs [hunk 2]
```diff
@@ -148,7 +155,29 @@ impl Collection {
         config.validate_and_warn();
         let configured_shards = config.params.shard_number.get();
 
-        for shard_id in 0..configured_shards {
+        let shard_ids_list: Vec<_> = match config.params.sharding_method.unwrap_or_default() {
+            ShardingMethod::Auto => (0..configured_shards).collect(),
```
[the remark this answers is not in the document]
Comment (generall) [reply lib/collection/src/collection/snapshots.rs:159]: `Fixed` assume we won't be able to dynamically scale later. `AutoIncrement` also not really accurate, cause we don't increment anything. I would rather call it `Even` or something like this. But the idea of `Auto` is to hide implementation details actually

File: lib/storage/src/content_manager/toc/create_collection.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (timvisee) [reply lib/storage/src/content_manager/toc/create_collection.rs:89]: Check! I was confused and thought this was the number assigned to a shard. But it's the total number of shards instead.

File: lib/collection/src/collection/snapshots.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (timvisee) [reply lib/collection/src/collection/snapshots.rs:159]: On second thought I agree that my suggestions aren't any better. Lets keep it at `Auto`.

File: lib/storage/src/content_manager/collection_meta_ops.rs [hunk 3]
```diff
@@ -104,11 +104,21 @@ pub struct CreateCollection {
...
-    /// Default is 1 for standalone, otherwise equal to the number of nodes
-    /// Minimum is 1
+    ///  - Default is 1 for standalone, otherwise equal to the number of nodes
+    ///  - Minimum is 1
+    /// For custom sharding:
+    /// Number of shards in collection per shard group.
+    ///  - Default is 1, meaning that each shard key will be mapped to a single shard
+    ///  - Minimum is 1
     #[serde(default)]
     pub shard_number: Option<u32>,
```
[the remark this answers is not in the document]
Comment (generall) [reply lib/storage/src/content_manager/collection_meta_ops.rs:116]: Those are already validated later, but for the sake of OpenAPI consistency I will update

[the remark this answers is not in the document]
Comment (generall) [reply lib/storage/src/content_manager/collection_meta_ops.rs:117]: :heavy_check_mark:

File: lib/collection/src/collection/snapshots.rs [hunk 4]
```diff
@@ -148,7 +155,29 @@ impl Collection {
...
+                );
+                if !mapping_path.exists() {
+                    Vec::new()
+                } else {
+                    let shard_key_mapping: ShardKeyMapping = read_json(&mapping_path)?;
+                    shard_key_mapping
+                        .values()
+                        .flat_map(|v| v.iter())
+                        .copied()
+                        .collect()
```
[the remark this answers is not in the document]
Comment (generall) [reply lib/collection/src/collection/snapshots.rs:175]: :heavy_check_mark:

**Review by timvisee — APPROVED** [review]

**Pull request merged**